mlflow/gateway/uc_function_utils.py

```diff
@@ -0,0 +1,309 @@
+                return {"type": "number"}
+            elif uc_type_json.startswith("interval"):
+                raise TypeError(f"Type {uc_type_json} is not supported.")
+            else:
+                raise TypeError(f"Unknown type {uc_type_json}. Try upgrading this package.")
+    else:
+        assert isinstance(uc_type_json, dict)
+        type = uc_type_json["type"]
+        if type == "array":
+            element_type = uc_type_to_json_schema_type(uc_type_json["elementType"])
+            return {"type": "array", "items": element_type}
+        elif type == "map":
+            key_type = uc_type_json["keyType"]
+            assert key_type == "string", TypeError(
+                f"Only support STRING key type for MAP but got {key_type}."
+            )
+            value_type = uc_type_to_json_schema_type(uc_type_json["valueType"])
+            return {
+                "type": "object",
+                "additionalProperties": value_type,
+            }
+        elif type == "struct":
+            properties = {}
+            for field in uc_type_json["fields"]:
+                properties[field["name"]] = uc_type_to_json_schema_type(field["type"])
+            return {"type": "object", "properties": properties}
+        else:
+            raise TypeError(f"Unknown type {uc_type_json}. Try upgrading this package.")
+
+
+def extract_param_metadata(p: FunctionParameterInfo) -> dict:
+    type_json = json.loads(p.type_json)["type"]
+    json_schema_type = uc_type_to_json_schema_type(type_json)
+    json_schema_type["name"] = p.name
+    json_schema_type["description"] = (
+        p.comment + f" (default: {p.parameter_default})" if p.parameter_default else ""
+    )
+    return json_schema_type
+
+
+def get_func_schema(func: "FunctionInfo") -> Dict[str, Any]:
+    return {
+        "description": func.comment,
+        "name": _get_tool_name(func),
+        "parameters": {
+            "type": "object",
+            "properties": {p.name: extract_param_metadata(p) for p in func.input_params.parameters},
+            "required": [
+                p.name for p in func.input_params.parameters if p.parameter_default is None
+            ],
+        },
+    }
+
+
+@dataclass
+class ParameterizedStatement:
+    statement: str
+    parameters: List["StatementParameterListItem"]
+
+
+@dataclass
+class FunctionExecutionResult:
+    """
+    Result of executing a function.
+    We always use a string to present the result value for AI model to consume.
+    """
+
+    error: Optional[str] = None
+    format: Optional[Literal["SCALAR", "CSV"]] = None
+    value: Optional[str] = None
+    truncated: Optional[bool] = None
+
+    def to_json(self) -> str:
+        data = {k: v for (k, v) in self.__dict__.items() if v is not None}
+        return json.dumps(data)
+
+
+def is_scalar(function: "FunctionInfo") -> bool:
+    from databricks.sdk.service.catalog import ColumnTypeName
+
+    return function.data_type != ColumnTypeName.TABLE_TYPE
```

Comment:
Is `is_scalar` the right name for this function? Here's the `ColumnTypeName` enum:

```
class ColumnTypeName(Enum):
    """Name of type (INT, STRUCT, MAP, etc.)."""

    ARRAY = 'ARRAY'
    BINARY = 'BINARY'
    BOOLEAN = 'BOOLEAN'
    BYTE = 'BYTE'
    CHAR = 'CHAR'
    DATE = 'DATE'
    DECIMAL = 'DECIMAL'
    DOUBLE = 'DOUBLE'
    FLOAT = 'FLOAT'
    INT = 'INT'
    INTERVAL = 'INTERVAL'
    LONG = 'LONG'
    MAP = 'MAP'
    NULL = 'NULL'
    SHORT = 'SHORT'
    STRING = 'STRING'
    STRUCT = 'STRUCT'
    TABLE_TYPE = 'TABLE_TYPE'
    TIMESTAMP = 'TIMESTAMP'
    TIMESTAMP_NTZ = 'TIMESTAMP_NTZ'
    USER_DEFINED_TYPE = 'USER_DEFINED_TYPE'
```

Things like `MAP` and `ARRAY` are also not scalar.

Reply:
Here 'scalar" means the return is a single row instead of a table.

Reply:
Got it. @harupy Can we add a docstring?